### PR TITLE
Feat: Add SET CREDIT + WIPE via Passthrough

### PR DESCRIPTION
**Pull request by jjmilburn**
## Contribution Guideline Acceptance

- [ X ] I have read and understand the [Individual Contributors License Agreement](CONTRIBUTING.md).
- [ X ] I accept the [Individual Contributors License Agreement](CONTRIBUTING.md) on the entire contents of this pull request.

## Proposed changes

Introduce 'Passthrough' commands in small protocol, and enable message obscuring (reordering/scrambling resulting digits) for passthrough messages.

Introduce the new "Set Credit + Wipe Restricted Flag" behavior through an 'extended small protocol', carried in the standard small protocol within a 'passthrough' message with 'app ID' = 1.

These messages do not carry a full 6-bit message ID, so additional logic has to be performed to remove the possibility of generating a message with a MAC that collides with another possible message in the same window.

## Types of changes

What types of changes does your code introduce to this project?
_Put an `x` in the boxes that apply_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [ X ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._

- [ X ] I have added unit tests with >= 90% coverage for all files modified or added by these changes
- [ X ] I have designed my tests to prove that my fix is effective or that my feature works as described
- [ X ] I have added docstrings to all new functions describing their purpose, parameters, and return value
- [X  ] I have updated or added supplemental documentation (if appropriate)
- [ ] Any dependent changes have been merged and published in downstream modules

## Further comments

If this is a relatively large or complex change, kick off the discussion by explaining why you chose the solution you did and what alternatives you considered, etc...

Link to any supplemental information here.

### Reviewers:

This will be filled by the people who review this particular Pull Request.

### Attribution

This PR template derived from:
https://github.com/skcript/PR-Template

Original template is:

Copyright (c) 2017 Skcript

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

### CODE REVIEW

File: nexus_keycode/protocols/small.py [hunk 1]
```diff
@@ -386,6 +412,192 @@ def __init__(self, type_, secret_key):
...
+                "Lowest ID with no chance of MAC collision is {} "
+                "(window {}-{}). Using ID {} instead of {}.".format(
+                    lowest_id,
+                    full_id,
+                    full_id + MESSAGE_ID_LOOKAHEAD_WINDOW,
+                    lowest_id,
+                    full_id)
+            )
+
+        if lowest_id == full_id + MESSAGE_ID_LOOKAHEAD_WINDOW:
```
Comment (jjmilburn) [inline nexus_keycode/protocols/small.py:581]: This logic likely won't be hit as-is.

File: nexus_keycode/protocols/small.py [hunk 2]
```diff
@@ -386,6 +412,192 @@ def __init__(self, type_, secret_key):
...
+
+    @classmethod
+    def generate_set_credit_wipe_restricted_flag_body(cls, id_, days):
+        # type: (int, int) -> bitstring.Bits
+        set_credit_increment_id = SetCreditSmallMessage.generate_body(days)
+
+        body_bits = bitstring.pack(
+            ["uint:2=truncated_msg_id", "uintle:8=increment_id"],
+            # truncate to least-significant 4 bits of the full ID
+            truncated_msg_id=id_ & 0b11,
```
Comment (jjmilburn) [inline nexus_keycode/protocols/small.py:594]: Todo: Use `SET_CREDIT_WIPE_RESTRICTED_FLAG.value[1]`

File: nexus_keycode/test/protocols/test_small.py [hunk 3]
```diff
@@ -214,6 +215,215 @@ def test_compressed_message_bits__unlock__output_correct(self):
...
+
+        computed_auth = protocol.ExtendedSmallMessage._compute_auth(
+            102,
+            protocol.ExtendedSmallMessageType.SET_CREDIT_WIPE_RESTRICTED_FLAG,
+            body_bits[4:14],  # body of inner 'extended' message
+            secret_key)
+
+        self.assertEqual(computed_auth, body_bits[14:26].uint)
+
+        # fixed MAC (matches 223 345 below)
```
Comment (ricehornet) [inline nexus_keycode/test/protocols/test_small.py:295]: This changed? Comment no longer makes sense.

File: nexus_keycode/test/protocols/test_small.py [hunk 4]
```diff
@@ -214,6 +215,215 @@ def test_compressed_message_bits__unlock__output_correct(self):
...
+        self.assertEqual(0b111100001010, body_bits[14:26].uint)
+        self.assertEqual("132 223 222 552 244", message.to_keycode())
+
+    def test_generate_set_credit_wipe_restricted__mac_collision__final_message_id_updated(self):
+        # expect collision, ID updated to '6'
```
Comment (ricehornet) [inline nexus_keycode/test/protocols/test_small.py:300]: Outdated comment?

**Review by ricehornet — COMMENTED** [review]
@jjmilburn this looks good to me -- just a few nits and suggestions for a few more checks. Leave it to @chadnorvell for final +1.

File: nexus_keycode/protocols/small.py [hunk 5]
```diff
@@ -1,25 +1,31 @@
...
 import math
 
 import bitstring
 import siphash
+from typing import Optional
 
 from nexus_keycode.protocols.utils import ints_to_bytes, pseudorandom_bits
 
-NEXUS_MODULE_VERSION_STRING = "1.1.0"
+NEXUS_MODULE_VERSION_STRING = "1.2.0"
```
[the remark this answers is not in the document]
Comment (jjmilburn) [reply nexus_keycode/protocols/small.py:12]: So, there are two different versions. The protocol subversions, and the "NEXUS_GLOBAL_VERSION". They should not be identical.

File: CHANGELOG.md [hunk 6]
```diff
@@ -10,3 +10,20 @@ Version 1.1.0 *(2020-07-31)*
 
 The Nexus Keycode encoder library now supports both Python 2.7 as well as 3.6
 and later.
+
+Version 2.2.0 *(2021-03-23)*
```
[the remark this answers is not in the document]
Comment (jjmilburn) [reply CHANGELOG.md:14]: Will cherry pick to bring it back in.

[the remark this answers is not in the document]
Comment (jjmilburn) [reply CHANGELOG.md:14]: Hmm, there was not a changelog for that tag. Do you see one?

File: README.md [hunk 7]
```diff
@@ -103,13 +103,6 @@ AddCreditSmallMessage(id_=42, days=7, secret_key=secret_key).to_keycode()
 # outputs 135 242 422 455 244
 ```
 
-Update Credit
-
-```python
```
Comment (jjmilburn) [inline README.md:108]: This was incorrect - not supported even in previous version.

File: nexus_keycode/protocols/small.py (hunk 2 above)
Comment (jjmilburn) [reply nexus_keycode/protocols/small.py:598]: Nevermind, should be 0b11 (value 3), not 2 (bit counter).

File: nexus_keycode/test/protocols/test_small.py (hunk 4 above)
Comment (jjmilburn) [reply nexus_keycode/test/protocols/test_small.py:300]: addressing

File: nexus_keycode/test/protocols/test_small.py (hunk 3 above)
Comment (jjmilburn) [reply nexus_keycode/test/protocols/test_small.py:295]: addressing

File: CHANGELOG.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (chadnorvell) [reply CHANGELOG.md:14]: The release is [here](https://github.com/angaza/nexus-python/releases/tag/1.2.0). I suspect we might not have changed the changelog. We can fix that now.

File: README.md (hunk 7 above)
Comment (chadnorvell) [reply README.md:108]: lol

Comment (chadnorvell) [reply README.md:108]: It was supported way back in 1.0, but we forgot to remove this example when we removed "update credit".

File: CHANGELOG.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (jjmilburn) [reply CHANGELOG.md:33]: I've updated the `CHANGELOG` to include the 1.2.0 release info and pushed!

**Pull request merged**